File: golang/3164_number_of_pairs/number_of_pairs.go
// Package main ...
package main

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func numberOfPairs(nums1 []int, nums2 []int, k int) int64 {
	count := make(map[int]int)
	count2 := make(map[int]int)
	max1 := 0
	for _, num := range nums1 {
		count[num]++
		if num > max1 {
			max1 = num
		}
	}
	for _, num := range nums2 {
		count2[num]++
	}
	var res int64
	for a, cnt := range count2 {
		for b := a * k; b <= max1; b += a * k {
			if _, ok := count[b]; ok {
				res += int64(count[b] * cnt)
			}
		}
	}
	return res
}

func main() {
	tests := []struct {
		nums1 []int
		nums2 []int
		k     int
		ans   int64
	}{
		{[]int{1, 3, 4}, []int{1, 3, 4}, 1, 5},
		{[]int{1, 2, 4, 12}, []int{2, 4}, 3, 2},
	}

	for index, test := range tests {
		assert.Equal(&testing.T{}, test.ans, numberOfPairs(test.nums1, test.nums2, test.k), index)
	}
}

File: cpp/3164_number_of_pairs/number_of_pairs.cpp
#include <cassert>
#include <tuple>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
 public:
  long long numberOfPairs(vector<int>& nums1, vector<int>& nums2, int k) {
    unordered_map<int, int> count, count2;
    int max1 = 0;
    for (int num : nums1) {
      count[num]++;
      max1 = max(max1, num);
    }
    for (int num : nums2) {
      count2[num]++;
    }
    long long res = 0;
    for (const auto& pair : count2) {
      int a = pair.first, cnt = pair.second;
      for (int b = a * k; b <= max1; b += a * k) {
        if (count.count(b) > 0) {
          res += 1L * count[b] * cnt;
        }
      }
    }
    return res;
  }
};

int main() {
  vector<tuple<vector<int>, vector<int>, int, long long>> tests{
      {{1, 3, 4}, {1, 3, 4}, 1, 5},
      {{1, 2, 4, 12}, {2, 4}, 3, 2},
  };

  for (auto &[nums1, nums2, k, ans] : tests) {
    assert(Solution().numberOfPairs(nums1, nums2, k) == ans);
  }
}
